src/zjs_pme.c

```diff
@@ -0,0 +1,393 @@
+        if (!zjs_pme_ipm_send_sync(&send, &reply)) {                       \
+            return zjs_error("ipm message failed or timed out!");          \
+        }                                                                  \
+        if (reply.error_code != ERROR_IPM_NONE) {                          \
+            ERR_PRINT("error code: %u\n", (unsigned int)reply.error_code); \
+            return zjs_error("error received");                            \
+        }                                                                  \
+    })
+
+static bool zjs_pme_ipm_send_sync(zjs_ipm_message_t *send,
+                                  zjs_ipm_message_t *result) {
+    send->id = MSG_ID_PME;
+    send->flags = 0 | MSG_SYNC_FLAG;
+    send->user_data = (void *)result;
+    send->error_code = ERROR_IPM_NONE;
+
+    if (zjs_ipm_send(MSG_ID_PME, send) != 0) {
+        ERR_PRINT("failed to send message\n");
+        return false;
+    }
+
+    // block until reply or timeout, we shouldn't see the ARC
+    // time out, if the ARC response comes back after it
+    // times out, it could pollute the result on the stack
+    if (k_sem_take(&pme_sem, ZJS_PME_TIMEOUT_TICKS)) {
+        ERR_PRINT("FATAL ERROR, ipm timed out\n");
+        return false;
+    }
+
+    return true;
+}
+
+static void ipm_msg_receive_callback(void *context, uint32_t id, volatile void *data)
+{
+    if (id != MSG_ID_PME)
+        return;
+
+    zjs_ipm_message_t *msg = (zjs_ipm_message_t *)(*(uintptr_t *)data);
+
+    if ((msg->flags & MSG_SYNC_FLAG) == MSG_SYNC_FLAG) {
+         zjs_ipm_message_t *result = (zjs_ipm_message_t *)msg->user_data;
+        // synchrounus ipm, copy the results
+        if (result)
+            memcpy(result, msg, sizeof(zjs_ipm_message_t));
+
+        // un-block sync api
+        k_sem_give(&pme_sem);
+    } else {
+        // asynchronous ipm, should not get here
+        ERR_PRINT("async message received\n");
+    }
+}
+
+static ZJS_DECL_FUNC(zjs_pme_begin)
+{
+    zjs_ipm_message_t send, reply;
+    send.type = TYPE_PME_BEGIN;
+
+    CALL_REMOTE_FUNCTION(send, reply);
+    return ZJS_UNDEFINED;
+}
+
+static ZJS_DECL_FUNC(zjs_pme_forget)
+{
+    zjs_ipm_message_t send, reply;
+    send.type = TYPE_PME_FORGET;
+
+    CALL_REMOTE_FUNCTION(send, reply);
+    return ZJS_UNDEFINED;
+}
+
+static ZJS_DECL_FUNC(zjs_pme_configure)
+{
+    ZJS_VALIDATE_ARGS(Z_NUMBER, Z_NUMBER, Z_NUMBER, Z_NUMBER, Z_NUMBER);
+    zjs_ipm_message_t send, reply;
+    send.type = TYPE_PME_CONFIGURE;
+
+    send.data.pme.g_context = jerry_get_number_value(argv[0]);
+    send.data.pme.c_mode = jerry_get_number_value(argv[1]);
+    send.data.pme.d_mode = jerry_get_number_value(argv[2]);
+    send.data.pme.min_influence = jerry_get_number_value(argv[3]);
+    send.data.pme.max_influence = jerry_get_number_value(argv[4]);
+    CALL_REMOTE_FUNCTION(send, reply);
+    return ZJS_UNDEFINED;
+}
+
+static ZJS_DECL_FUNC(zjs_pme_learn)
+{
+    ZJS_VALIDATE_ARGS(Z_ARRAY, Z_NUMBER);
```

Comment:
add a comment before every ZJS_VALIDATE_ARGS describing the expected args (e.g. names: array of patterns, count or whatever)